src/newwave/bridge.mo

```diff
@@ -0,0 +1,235 @@
+import Time "mo:base/Time";
+import Int "mo:base/Int";
+import Nat64 "mo:base/Nat64";
+
+import Error "mo:base/Error";
+import Hash "mo:base/Hash";
+import HashMap "mo:base/HashMap";
+import Nat "mo:base/Nat";
+import Option "mo:base/Option";
+import Principal "mo:base/Principal";
+import Array "mo:base/Array";
+import Iter "mo:base/Iter";
+import P "mo:base/Prelude";
+import Nat32 "mo:base/Nat32";
+import Result "mo:base/Result";
+
+import Random "mo:base/Random";
+import Blob "mo:base/Blob";
+import Utils "Utils";
+import Types "Types";
+
+module {
+
+  /**
+   * Defines the errors for the public API when trying to retrieve a Bridge ID
+  */
+  public type BridgeIdErrors = {
+    #Unauthorized;
+    #BridgeNotFound;
+    #Error;
+  };
+
+  /**
+   * Defines the result type for trying to retieve the ID of a bridge from
+   * the public API
+  */
+  public type BridgeIdResult = Types.Result<Text, BridgeIdErrors>;
+
+  /**
+   * Defines the bridge errors that can occur when trying to retrieve a bridge
+  */
+  public type BridgeErrors = {
+    #Unauthorized : Text;
+    #BridgeNotFound;
+    #Error;
+  };
+
+  /**
+   * Defines the result type for trying to retrieve an Entity from
+   * the public API
+  */
+  public type BridgeResult = Types.Result<Bridge, BridgeErrors>;
+
+  /**
+   * Stores bridge specific settings
+  */
+  public class BridgeSettings() {
+    var mainSetting : Text = "default";
+  };
+
+  /**
+   * The available bridge types that can be used to describe a bridge
+  */
+  public type BridgeType = {
+    #OwnerCreated;
+  };
+
+  /**
+    * The bridge states that describe the current status of a bridge
+  */
+  public type BridgeState = {
+    #Pending;
+    #Rejected;
+    #Confirmed;
+  };
+
+  /**
+    * The type that defines the attributes for a Bridge
+  */
+  public type Bridge = {
+    /**
+     * The ID of the Bridge that is used to store it in
+     * in the bridge database
+    */
+    id : Text;
+    /**
+     * The timestamp in UTC (maybe) that the bridge was created
+    */
+    creationTimestamp : Nat64;
+    /**
+     * The original creator of the bridge.
+    */
+    creator : Principal;
+    /**
+     * The current owner of the bridge
+    */
+    owner : Principal;
+    /**
+     * Settings for the bridge
+    */
+    settings : BridgeSettings;
+    /**
+     * A human readable name? for the bridge
+    */
+    name : ?Text;
+    /**
+     * An owner defined description for what the bridge is
+    */
+    description : ?Text;
+    /**
+     * Keywords that are used to descripe the bridge to
+     * enable more efficient lookup of the bridge?
+    */
+    keywords : ?[Text];
+    /**
+     * Unknown
+    */
+    listOfBridgeSpecificFieldKeys : [Text];
```

Comment:
we can move the upper attributes (id through keywords) into the BaseEntity type and delete listOfBridgeSpecificFieldKeys (the BaseEntity type likely has an attribute listOfEntitySpecificFieldKeys or listOfTypeSpecificFieldKeys)